File: frontend/lib/schemas/v2/patient_fixed_visit.ts
/**
 * PatientFixedVisitV2 zod schemas (W9-FE1 / W22 / W37 Phase 3-A).
 *
 * 設計仕様書 `docs/plans/v2-allocation-redesign.md` §3.6.8 (固定枠ライフサイクル)
 * および `docs/plans/v2-api-contracts.md` §8 / §9 に基づく型定義。
 *
 * Backend `backend/app/schemas/v2/patient_fixed_visit.py` と完全一致させる。
 *
 * W22: course_template_id (UUID | null) を追加。
 *   BE Wave 22 (feat/v2-w22-pfv-course) で migration 0025 として追加された列。
 *   並列実装のため型契約を先行ミラー。BE 完成後に再同期する。
 *
 * W37 Phase 1 / Phase 3-A:
 *   `slot_index` (0 or 1) を追加。複数スタッフ対応患者は同曜日・同時刻に
 *   2 コース固定するため slot_index 0/1 の 2 行を持つ。1 名体制では default 0.
 *   UNIQUE 制約は (patient_id, mode, weekday, slot_index)。FE bulk PUT も
 *   (weekday, slot_index) ペアで重複検出する。
 */
import { z } from 'zod';

export const PATIENT_FIXED_VISIT_MODES = ['normal', 'special'] as const;
export type PatientFixedVisitMode = (typeof PATIENT_FIXED_VISIT_MODES)[number];

/**
 * W37 Phase 1: スロット番号 (0 or 1).
 * 1 名体制 (既存) では 0 のみ。
 * 複数スタッフ対応では同 weekday に 0/1 の 2 行を持つ。
 */
export const SLOT_INDEX_VALUES = [0, 1] as const;
export type SlotIndex = (typeof SLOT_INDEX_VALUES)[number];

/**
 * P2-A: 可動域フラグ = 提案の可否 (改善提案 MVP).
 * unknown(既定・保守的) / time_flexible(同曜日内の時刻変更可) /
 * day_flexible(曜日も変更可) / locked(完全固定).
 */
export const MOVABILITY_VALUES = ['unknown', 'time_flexible', 'day_flexible', 'locked'] as const;
export type Movability = (typeof MOVABILITY_VALUES)[number];

export const patientFixedVisitV2BaseSchema = z.object({
  weekday: z.number().int().min(0).max(6),
  start_time: z.string().regex(/^([01]\d|2[0-3]):[0-5]\d(?::[0-5]\d)?$/, 'HH:MM 形式'),
  // 基本の訪問時間 35 分をデフォルトに (PO 決定 2026-08-09。BE 側デフォルトと同値)。
  duration_min: z.number().int().min(1).max(480).default(35),
  /** W22: コーステンプレート ID (UUID | null). null = 未指定 (Layer 1 フォールバック). */
  course_template_id: z.string().uuid().nullable().optional(),
  /**
   * W37 Phase 1: 同 weekday × 同 mode 内のスロット番号 (0 or 1).
   * 1 名体制では 0 のみ送信し、BE 側 default も 0。
   * 複数スタッフ対応 (requires_multiple_staff=true) では 0/1 の 2 行を送る。
   * 後方互換: 省略時は BE 側で default=0 が付く。
   */
  slot_index: z
    .number()
    .int()
    .min(0)
    .max(1)
    .default(0)
    .describe(
      '同 patient × 同 mode × 同 weekday の中のスロット番号 (0 or 1). 1 名体制では 0 のみ.',
    ),
  /**
   * Phase E-5 (項目 ⑥B): サブ拠点 ID.
   * 主担当拠点 (patient.primary_office_id) とは別に、フォロー時の配置先を指定する.
   * null/undefined = 未指定 (従来挙動). 指定時は course_template_id も sub_office の
   * template を指す必要がある (BE 側で 422 検証).
   */
  sub_office_id: z.string().uuid().nullable().optional(),
  /**
   * Phase G-21: 「完全固定」フラグ.
   * true のとき、自動最適化 (Layer 2) はこの PFV から生成された visit を移動・削除しない.
   * Read 用途では BE が値を返す. Write 用途 (bulk PUT) では省略可 (BE は既存値維持 or false default).
   */
  is_pinned: z.boolean().optional(),
  /**
   * P2-A: 可動域フラグ = 提案の可否.
   * is_pinned と同じく `.optional()` (default なし). 理由:
   *   - Read/parse: movability を返さない旧 BE レスポンスでも `undefined` で壊れない
   *     (BE 先行デプロイ互換). 送らない旧 FE リクエストも BE 側 server_default 'unknown'
   *     で従来挙動を維持する (既定挙動不変).
   *   - Write/build: proposedSlotToFixedVisitItem (新規枠=unknown が正) を変更せずに済む
   *     よう、build 側で省略可能にする (`.default()` は infer 出力型を必須化し builder を
   *     壊すため使わない — is_pinned と同一方針).
   */
  movability: z.enum(MOVABILITY_VALUES).optional(),
});

export const patientFixedVisitV2ReadSchema = patientFixedVisitV2BaseSchema.extend({
  id: z.string().uuid(),
  patient_id: z.string().uuid(),
  mode: z.enum(PATIENT_FIXED_VISIT_MODES),
  created_at: z.string(),
  updated_at: z.string(),
});

/**
 * W37 Phase 1: bulk PUT は (weekday, slot_index) のペアで重複を検出する。
 * 1 名体制では従来どおり weekday ごとに 1 件のみだが、
 * 複数スタッフ対応では同じ weekday で slot_index 0/1 の 2 件を許容する。
 *
 * U-1 (変更反映先統一): change_scope / iso_year / iso_week を追加。
 *   - change_scope='pattern_only' (既定・従来): 固定訪問週間（型）のみ変更。
 *   - change_scope='pattern_and_week': 型を変え、今週のスケジュールにも即反映する
 *     (sync-fixed-to-week を 1 TX 後段実行)。iso_year / iso_week は必須。
 *   旧 BE 互換: フィールド省略時は BE 側 default (pattern_only = 従来挙動) が適用される。
 */
export const patientFixedVisitsBulkPutSchema = z.object({
  mode: z.enum(PATIENT_FIXED_VISIT_MODES),
  items: z
    .array(patientFixedVisitV2BaseSchema)
    // 7 曜日 × slot 0/1 = 最大 14 件
    .max(14)
    .superRefine((items, ctx) => {
      const seen = new Map<string, number>();
      items.forEach((it, i) => {
        // slot_index が省略された場合は default 0 として扱う
        const slot = it.slot_index ?? 0;
        const key = `${it.weekday}-${slot}`;
        if (seen.has(key)) {
          ctx.addIssue({
            code: z.ZodIssueCode.custom,
            path: [i, 'weekday'],
            message: '同じ曜日・スロットが重複しています',
          });
        }
        seen.set(key, i);
      });
    }),
  /**
   * U-1: 反映先選択。省略時は BE 側 default (pattern_only = 従来挙動)。
   * 'pattern_and_week' 選択時は iso_year / iso_week も送る。
   */
  change_scope: z.enum(['pattern_only', 'pattern_and_week']).optional(),
  /** U-1: change_scope='pattern_and_week' のときの対象 ISO 年. */
  iso_year: z.number().int().optional(),
  /** U-1: change_scope='pattern_and_week' のときの対象 ISO 週. */
  iso_week: z.number().int().min(1).max(53).optional(),
  // 方式b: 定員超過採用時の管理者判断理由 (最大500字)。通常候補では null/省略。旧BE互換。
  capacity_override_reason: z.string().max(500).nullish(),
});

export type PatientFixedVisitV2Base = z.infer<typeof patientFixedVisitV2BaseSchema>;
export type PatientFixedVisitV2Read = z.infer<typeof patientFixedVisitV2ReadSchema>;
export type PatientFixedVisitsBulkPut = z.infer<typeof patientFixedVisitsBulkPutSchema>;

/**
 * P0-2 Commit 1 (BE `731884d`): PUT /patients/{id}/fixed-visits のレスポンスが
 * `{items, warnings}` エンベロープに変更された。warnings 要素は再検証カーネル
 * (`pfv_validator.py`) の `PfvValidationWarningOut` = {code, message, weekday, severity}。
 *
 * 寛容パース方針: code / severity は未知値を壊さないよう `z.string()` (enum 化しない)。
 * weekday は nullable (曜日非依存の警告は null)。将来 BE が新フィールドを足しても
 * FE 採用フローを止めないよう `.passthrough()` で余剰キーを許容する。
 */
export const patientFixedVisitWarningSchema = z
  .object({
    // 未知 code を壊さない寛容パース (enum 化しない).
    code: z.string(),
    message: z.string(),
    // 曜日非依存の警告は null / 省略あり.
    weekday: z.number().int().min(0).max(6).nullable().optional(),
    // 'warning' | 'error' 等. 未知 severity も string で受ける.
    severity: z.string(),
  })
  .passthrough();
export type PatientFixedVisitWarning = z.infer<typeof patientFixedVisitWarningSchema>;

// ─── 案Z (PO 決定 2026-08-09): dry-run 検証 + コース負荷 ─────────────────────

/** POST /fixed-visits/validate のレスポンス (保存せず警告だけ返す)。 */
export const pfvValidateResponseSchema = z.object({
  warnings: z.array(patientFixedVisitWarningSchema).catch([]),
});
export type PfvValidateResponse = z.infer<typeof pfvValidateResponseSchema>;

/** GET /fixed-visits/course-load — (曜日×コース) の他患者負荷。空き表示の材料。 */
export const pfvCourseLoadCellSchema = z.object({
  weekday: z.number().int(),
  course_template_id: z.string().uuid().nullable(),
  used_minutes: z.number().int(),
  patient_count: z.number().int(),
});
export const pfvCourseLoadResponseSchema = z.object({
  course_max_minutes: z.number().int(),
  max_patients_per_course: z.number().int(),
  cells: z.array(pfvCourseLoadCellSchema).catch([]),
});
export type PfvCourseLoadResponse = z.infer<typeof pfvCourseLoadResponseSchema>;

/**
 * PUT /patients/{id}/fixed-visits のエンベロープレスポンス。
 * - items: 既存 Read スキーマで検証。要素 drift 時は `.catch([])` で items のみ空に落とし、
 *   warnings 表示 (本 Commit の主目的) を巻き込まない。
 * - warnings: 上記の寛容な warning スキーマ配列。
 *
 * 呼出側 (hook) は `.safeParse` でさらに包み、object でない等の致命ケースでも
 * `{items:[], warnings:[]}` にフォールバックして採用フローを止めない。
 */
export const patientFixedVisitsBulkPutResponseSchema = z.object({
  items: z.array(patientFixedVisitV2ReadSchema).catch([]),
  warnings: z.array(patientFixedVisitWarningSchema).catch([]),
  // Wave U-1: change_scope=pattern_and_week のとき今週再生成の件数。
  // 旧 BE では欠落 / 週再生成に失敗した場合は null (FE は「週を生成」再実行を案内する)。
  week_sync: z
    .object({
      visits_regenerated: z.number().int(),
      visits_soft_deleted: z.number().int(),
    })
    .nullable()
    .optional()
    .default(null)
    .catch(null),
});
export type PatientFixedVisitsBulkPutResponse = z.infer<
  typeof patientFixedVisitsBulkPutResponseSchema
>;

/**
 * PUT fixed-visits レスポンスの寛容パース。parse 失敗 (object でない / null 等) でも
 * throw せず `{items:[], warnings:[]}` を返し、採用・保存フローを止めない。
 */
export function parsePatientFixedVisitsBulkPutResponse(
  raw: unknown,
): PatientFixedVisitsBulkPutResponse {
  const parsed = patientFixedVisitsBulkPutResponseSchema.safeParse(raw);
  return parsed.success ? parsed.data : { items: [], warnings: [], week_sync: null };
}

File: backend/app/schemas/v2/patient_fixed_visit.py
"""Pydantic schemas for patient_fixed_visits (W9-BE1 / W37 Phase 1).

固定訪問パターン の入力・出力型定義。
フロントエンド zod schema と完全一致を目指す。

W37 Phase 1: 複数スタッフ対応患者は同曜日・同時刻に 2 コース固定するため
slot_index (0/1) を導入。1 名体制では default 0 のみ使う。
"""

from __future__ import annotations

from datetime import datetime, time
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

PatientFixedVisitMode = Literal["normal", "special"]

# P2-A: 可動域フラグ = 提案の可否 (改善提案 MVP).
# unknown(既定・保守的) / time_flexible(同曜日内の時刻変更可) /
# day_flexible(曜日も変更可) / locked(完全固定).
PatientFixedVisitMovability = Literal["unknown", "time_flexible", "day_flexible", "locked"]


class PatientFixedVisitV2Base(BaseModel):
    """PUT body / 個別訪問アイテムの共通フィールド."""

    model_config = ConfigDict(extra="forbid")

    weekday: int = Field(ge=0, le=6, description="0=月 … 6=日")
    start_time: time = Field(description="訪問開始時刻 HH:MM")
    # 基本の訪問時間 35 分をデフォルトに (PO 決定 2026-08-09。FE zod と同値)。
    duration_min: int = Field(ge=1, le=480, default=35, description="訪問時間 (分)")
    # W22 Phase A: 固定枠が属するコーステンプレート ID (省略可).
    course_template_id: UUID | None = Field(
        default=None,
        description=(
            "固定枠が属する course_templates.id (W22). "
            "未指定なら Layer 1 の office フォールバックで解決される。"
        ),
    )
    # Phase E-5 (項目 ⑥B): サブ拠点 ID (省略可).
    # 主担当 (patient.primary_office_id) とは別にフォロー時に PFV を配置できる拠点.
    # 指定時は course_template_id も sub_office の template を指す必要がある (API で検証).
    sub_office_id: UUID | None = Field(
        default=None,
        description=(
            "サブ拠点 ID (Phase E-5 ⑥B). 未指定なら従来どおり主担当拠点のみで動作. "
            "指定時は course_template_id も sub_office の template を指すこと."
        ),
    )
    # W37 Phase 1: 同曜日・同時刻に 2 コース固定するためのスロット番号.
    # 1 名体制 (既存) では 0 のみ. 複数スタッフ対応では 0/1 の 2 行を持つ.
    slot_index: int = Field(
        default=0,
        ge=0,
        le=1,
        description=(
            "同 patient × 同 mode × 同 weekday の中のスロット番号 (0 or 1). "
            "1 名体制では 0 のみ. 複数スタッフ対応 (W37) で 0/1 の 2 行を持つ."
        ),
    )
    # Phase G-21: 完全固定フラグ. true = 自動算出で絶対動かさない.
    # 既存全 PFV は migration で true backfill 済 (後方互換).
    # 新規 PFV は default false.
    is_pinned: bool = Field(
        default=False,
        description=(
            "Phase G-21: 完全固定フラグ. true = 自動算出で絶対動かさない. "
            "既存全 PFV は migration で true backfill 済 (後方互換)."
        ),
    )
    # P2-A: 可動域フラグ = 提案の可否 (改善提案 MVP).
    # 送らない旧 FE リクエストは default 'unknown' で保存され従来と同一動作 (既定挙動不変).
    # is_pinned=True かつ movability != 'locked' は pfv_validator V6 が 'locked' に矯正する.
    movability: PatientFixedVisitMovability = Field(
        default="unknown",
        description=(
            "P2-A: 可動域フラグ = 提案の可否. "
            "unknown(既定) / time_flexible / day_flexible / locked. "
            "未指定は 'unknown' で保存 (従来挙動)."
        ),
    )


class PatientFixedVisitV2Read(PatientFixedVisitV2Base):
    """GET レスポンスの単一行."""

    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: UUID
    patient_id: UUID
    mode: PatientFixedVisitMode
    created_at: datetime
    updated_at: datetime


class PatientFixedVisitsBulkPut(BaseModel):
    """PUT /patients/{id}/fixed-visits の一括上書きリクエスト body.

    items 内で (weekday, slot_index) が重複している場合は 422 を返す。
    W37 Phase 1: 1 名体制では従来どおり weekday ごとに 1 件のみ (slot_index=0).
                 複数スタッフ対応では同 weekday に slot_index 0/1 の 2 件まで許容.
    """

    model_config = ConfigDict(extra="forbid")

    mode: PatientFixedVisitMode
    items: list[PatientFixedVisitV2Base] = Field(
        default_factory=list,
        # W37 Phase 1: 7 曜日 × slot 0/1 = 最大 14 件.
        max_length=14,
        description=("0〜14 件 (7 曜日 × slot 0/1)。(weekday, slot_index) 重複は不可。"),
    )
    # Wave U-1 (§2.2 反映先の統一 / プール採用 A 経路):
    #   pattern_only      = 型 (PFV) のみ上書き (既定・従来挙動; 後方互換).
    #   pattern_and_week  = PFV 上書き後、同一リクエスト内で当該患者の今週 visits を
    #                       PFV から再生成する (= sync-fixed-to-week 相当).
    change_scope: Literal["pattern_only", "pattern_and_week"] = Field(
        default="pattern_only",
        description=(
            "反映先の統一 (Wave U-1). pattern_only=型のみ(既定) / "
            "pattern_and_week=型+今週再生成(A). 後者は iso_year/iso_week 必須."
        ),
    )
    # pattern_and_week のとき再生成する対象週 (必須). pattern_only では無視.
    iso_year: int | None = Field(default=None, ge=2020, le=2100)
    iso_week: int | None = Field(default=None, ge=1, le=53)

    # 定員超過の管理者相談プロセス (方式b): 定員超過を承知で採用したときの理由 (監査用).
    # 値があれば logger.info で記録するだけで、検証や挙動変更は一切しない
    # (pfv_validator の容量チェックは既に warning-only). 後方互換の任意項目.
    capacity_override_reason: str | None = Field(default=None, max_length=500)

    # NG スタッフ / 性別制限の確認フロー (docs/plans/patient-ng-staff-design.md §7-2)。
    # 既定 False = 従来どおり。change_scope='pattern_and_week' のときのみ検査対象
    # (週が特定できて初めて「どのコースの誰に当たるか」が決まる)。
    acknowledge_constraint_warnings: bool = Field(
        default=False,
        description="NGスタッフ/性別制限の警告を確認済みとして続行する (422 の再送用)。",
    )

    @model_validator(mode="after")
    def _no_duplicate_weekday_slot(self) -> PatientFixedVisitsBulkPut:
        keys = [(item.weekday, item.slot_index) for item in self.items]
        if len(keys) != len(set(keys)):
            raise ValueError("items に (weekday, slot_index) の重複があります")
        return self


# ---------------------------------------------------------------------------
# P0-2: PUT レスポンスのエンベロープ化 (再検証警告の同梱)
# ---------------------------------------------------------------------------


class PfvValidationWarningOut(BaseModel):
    """PUT /fixed-visits の再検証で検出した 1 件の警告.

    ``severity="warning"`` のみをレスポンスに載せる (error は 422 で返す).
    code は pfv_validator の 4 種 (patient_time_conflict / lunch_break_overlap /
    course_capacity_exceeded / pinned_protection).
    """

    model_config = ConfigDict(extra="forbid")

    code: str = Field(description="警告種別コード (pfv_validator 準拠).")
    message: str = Field(description="現場向け日本語メッセージ.")
    weekday: int = Field(ge=0, le=6, description="0=月 … 6=日.")
    severity: str = Field(description='"warning" (続行可) / "error" (422 対象).')


class PfvWeekSyncOut(BaseModel):
    """PUT /fixed-visits の change_scope=pattern_and_week で今週 visits を再生成した件数.

    Wave U-1 (§2.2 A 経路). pattern_only のときは ``week_sync=null`` (後方互換).
    """

    model_config = ConfigDict(extra="forbid")

    visits_regenerated: int = Field(ge=0, description="PFV から再生成した今週 visit 数")
    visits_soft_deleted: int = Field(ge=0, description="再生成前に soft-delete した今週 visit 数")


class PatientFixedVisitsBulkPutResponse(BaseModel):
    """PUT /patients/{id}/fixed-visits のエンベロープレスポンス (P0-2).

    ``items`` は従来どおり確定後の全 PFV 行. ``warnings`` は再検証で検出した
    非致命の指摘 (患者間衝突 / 昼休み / 容量). FE 未パースのため BE 先行デプロイ安全.

    Wave U-1: ``week_sync`` は change_scope=pattern_and_week のときのみ非 null
    (今週 visits 再生成の件数). 追加フィールドのみで後方互換.
    """

    model_config = ConfigDict(extra="forbid")

    items: list[PatientFixedVisitV2Read]
    warnings: list[PfvValidationWarningOut] = Field(default_factory=list)
    week_sync: PfvWeekSyncOut | None = Field(default=None)


# ---------------------------------------------------------------------------
# 案Z (PO 決定 2026-08-09): マスタ編集の「賢い化」用 dry-run 検証 + コース負荷
# ---------------------------------------------------------------------------


class PfvValidateRequest(BaseModel):
    """POST /patients/{id}/fixed-visits/validate — 保存せずに再検証だけ実行する。

    入力中のライブ検査用 (常設表示)。PUT と同じ items 形で、削除も INSERT もしない。
    """

    model_config = ConfigDict(extra="forbid")

    mode: PatientFixedVisitMode
    items: list[PatientFixedVisitV2Base] = Field(default_factory=list, max_length=14)


class PfvValidateResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    warnings: list[PfvValidationWarningOut] = Field(default_factory=list)


class PfvCourseLoadCell(BaseModel):
    """(曜日 × コース) 単位の既存負荷 (対象患者自身を除く他患者の型の合計)。"""

    model_config = ConfigDict(extra="forbid")

    weekday: int
    course_template_id: UUID | None
    used_minutes: int
    patient_count: int


class PfvCourseLoadResponse(BaseModel):
    """GET /patients/{id}/fixed-visits/course-load — コースセレクトの空き表示用。"""

    model_config = ConfigDict(extra="forbid")

    course_max_minutes: int
    max_patients_per_course: int
    cells: list[PfvCourseLoadCell] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Phase G-21 T2: is_pinned 単独切替用 schema
# ---------------------------------------------------------------------------


class PatientFixedVisitPinUpdate(BaseModel):
    """PATCH /patients/.../fixed-visits/{pfv_id}/pin の body.

    既存 PFV 行の ``is_pinned`` のみ単独切替する用途.
    """

    model_config = ConfigDict(extra="forbid")

    is_pinned: bool = Field(description="Phase G-21: true = 自動算出で絶対動かさない完全固定.")


class PatientFixedVisitPinBulkItem(BaseModel):
    """POST /patients/fixed-visits/pin/bulk の items 1 件.

    複数 PFV の is_pinned を一括切替する用途.
    """

    model_config = ConfigDict(extra="forbid")

    pfv_id: UUID = Field(description="対象 patient_fixed_visits.id")
    is_pinned: bool = Field(description="Phase G-21: true = 自動算出で絶対動かさない完全固定.")
